File: leetCode/fukun/lc_031.js
const nextPermutation = function(nums) {
    let maxIndex = -Infinity,
        changeIndex;
    //找到交换点
    for (let i = nums.length - 1; i >= 0; i--) {
        for (let j = i - 1; j >= 0; j--) {
            if (nums[i] > nums[j]) {
                if (maxIndex < j) {
                    maxIndex = j;
                    changeIndex = i;
                }
            }
        }
    }
    //交换后就再逆序
    if (changeIndex) {
        [nums[maxIndex], nums[changeIndex]] = [nums[changeIndex], nums[maxIndex]];//交换写法，可以参考。
        for (let i = maxIndex + 1; i < nums.length; i++) {
            let min = i;
            for (let j = i + 1; j < nums.length; j++) {
                if (nums[j] < nums[min]) {
                    min = j;
                }
            }
            [nums[i], nums[min]] = [nums[min], nums[i]];
        }
    } else {
        nums.reverse();
    }
};

File: leetCode/XSS/lc_031.java
import java.util.Arrays;

public class lc_031 {
    public static void main(String[] args) {
        int nums[]=new int[]{3,2,1,4};
        new Solution().nextPermutation(nums);
    }
    static class Solution {
        public void nextPermutation(int[] nums) {
            int end=nums.length-1;
            int a=-1;int b=-1;
            for(int i=end-1;i>=0;i--)
            {
                if(nums[i]<nums[i+1])
                {
                    a=i;
                    break;
                }
            }
            if(a!=-1)
            {
                for(int i=end;i>a;i--)
                {
                    if(nums[i]>nums[a])
                    {
                        b=i;
                        break;
                    }
                }
                int temp=nums[a];
                nums[a]=nums[b];
                nums[b]=temp;
                reverse(nums,a+1,nums.length-1);
            }
            else
            {
                reverse(nums,0,nums.length-1);
            }
            for(int i=0;i<nums.length;i++)
            {
            System.out.print(nums[i]);}

        }
        public void reverse(int n[],int start,int end)
        {
            int len=end-start;
            for(int i=0;i<=len/2;i++)
            {
                int temp=n[i+start];
                n[i+start]=n[end-i];
                n[end-i]=temp;
            }
        }
    }
}
